File: algorithms/sudoku/sudoku.cc
#include <array>
#include <bitset>
#include <cctype>
#include <iostream>
#include <optional>
#include <ostream>
#include <utility>

class Sudoku {
 public:
  int &operator[](std::pair<int, int> pos) {
    return this->board_[pos.first][pos.second];
  }

  [[nodiscard]] int operator[](std::pair<int, int> pos) const {
    return this->board_[pos.first][pos.second];
  }

  [[nodiscard]] bool IsValid() const {
    for (int i = 0; i < 9; ++i) {
      std::bitset<10> row_check = {};
      std::bitset<10> col_check = {};
      std::bitset<10> box_check = {};

      for (int j = 0; j < 9; ++j) {
        int row_value = (*this)[{i, j}];
        if (row_value != 0) {
          if (row_check[row_value]) {
            return false;
          }
          row_check[row_value] = true;
        }

        int col_value = (*this)[{j, i}];
        if (col_value != 0) {
          if (col_check[col_value]) {
            return false;
          }
          col_check[col_value] = true;
        }

        int box_row = ((i / 3) * 3) + (j / 3);
        int box_col = ((i % 3) * 3) + (j % 3);
        int box_value = (*this)[{box_row, box_col}];
        if (box_value != 0) {
          if (box_check[box_value]) {
            return false;
          }
          box_check[box_value] = true;
        }
      }
    }

    return true;
  }

  [[nodiscard]] bool IsSolved() const {
    for (int i = 0; i < 9; ++i) {
      for (int j = 0; j < 9; ++j) {
        if ((*this)[{i, j}] == 0) {
          return false;
        }
      }
    }
    return this->IsValid();
  }

  [[nodiscard]] std::optional<Sudoku> Solve() const {
    if (this->IsSolved()) {
      return *this;
    }
    if (!this->IsValid()) {
      return {};
    }
    Sudoku copy = *this;
    for (int i = 0; i < 9; ++i) {
      for (int j = 0; j < 9; ++j) {
        if (copy[{i, j}] == 0) {
          for (int num = 1; num <= 9; ++num) {
            copy[{i, j}] = num;
            std::optional<Sudoku> result = copy.Solve();
            if (result.has_value()) {
              return result;
            }
            copy[{i, j}] = 0;
          }
          return {};
        }
      }
    }
    return {};
  }

 private:
  std::array<std::array<int, 9>, 9> board_ = {};
};

std::ostream &operator<<(std::ostream &os, const Sudoku &sudoku) {
  os << "+-------+-------+-------+\n";
  for (int i = 0; i < 9; ++i) {
    os << "| ";
    for (int j = 0; j < 9; ++j) {
      int value = sudoku[{i, j}];
      if (value == 0) {
        os << ".";
      } else {
        os << value;
      }
      os << " ";
      if ((j + 1) % 3 == 0 && j < 8) {
        os << "| ";
      }
    }
    os << "|\n";
    if ((i + 1) % 3 == 0 && i < 8) {
      os << "+-------+-------+-------+\n";
    }
  }
  os << "+-------+-------+-------+\n";
  return os;
}

int main() {
  Sudoku sudoku;

  for (int i = 0; i < 9; ++i) {
    for (int j = 0; j < 9; ++j) {
      int value;
      for (;;) {
        value = std::cin.get();
        if (value == std::char_traits<char>::eof()) {
          std::cerr << "Unexpected EOF\n";
          return 1;
        }
        // NOLINTNEXTLINE(readability-implicit-bool-conversion)
        if (std::isspace(value)) {
          continue;
        }
        break;
      }
      // NOLINTNEXTLINE(readability-implicit-bool-conversion)
      if (std::isdigit(value)) {
        sudoku[{i, j}] = value - '0';
      }
    }
  }

  std::cout << "Original:\n" << sudoku << "\n";
  std::optional<Sudoku> result = sudoku.Solve();
  if (!result.has_value()) {
    std::cout << "No solution found\n";
    return 1;
  }
  std::cout << "Solved:\n" << *result << "\n";
}

File: algorithms/sudoku/sudoku.js
class Sudoku {
  constructor() {
    this.board = Array(9).fill().map(() => Array(9).fill(0));
  }

  clone() {
    const copy = new Sudoku();
    for (let i = 0; i < 9; ++i) {
      for (let j = 0; j < 9; ++j) {
        copy.board[i][j] = this.board[i][j];
      }
    }
    return copy;
  }

  getValue(pos) {
    return this.board[pos[0]][pos[1]];
  }

  setValue(pos, value) {
    this.board[pos[0]][pos[1]] = value;
    return value;
  }

  isValid() {
    for (let i = 0; i < 9; ++i) {
      const rowCheck = Array(10).fill(false);
      const colCheck = Array(10).fill(false);
      const boxCheck = Array(10).fill(false);

      for (let j = 0; j < 9; ++j) {
        const rowValue = this.getValue([i, j]);
        if (rowValue !== 0) {
          if (rowCheck[rowValue]) {
            return false;
          }
          rowCheck[rowValue] = true;
        }

        const colValue = this.getValue([j, i]);
        if (colValue !== 0) {
          if (colCheck[colValue]) {
            return false;
          }
          colCheck[colValue] = true;
        }

        const boxRow = Math.floor(i / 3) * 3 + Math.floor(j / 3);
        const boxCol = (i % 3) * 3 + (j % 3);
        const boxValue = this.getValue([boxRow, boxCol]);
        if (boxValue !== 0) {
          if (boxCheck[boxValue]) {
            return false;
          }
          boxCheck[boxValue] = true;
        }
      }
    }
    return true;
  }

  isSolved() {
    for (let i = 0; i < 9; ++i) {
      for (let j = 0; j < 9; ++j) {
        if (this.getValue([i, j]) === 0) {
          return false;
        }
      }
    }
    return this.isValid();
  }

  solve() {
    if (this.isSolved()) {
      return this;
    }
    if (!this.isValid()) {
      return null;
    }

    const copy = this.clone();
    for (let i = 0; i < 9; ++i) {
      for (let j = 0; j < 9; ++j) {
        if (copy.getValue([i, j]) === 0) {
          for (let num = 1; num <= 9; ++num) {
            copy.setValue([i, j], num);
            const result = copy.solve();
            if (result !== null) {
              return result;
            }
            copy.setValue([i, j], 0);
          }
          return null;
        }
      }
    }
    return null;
  }

  toString() {
    let result = "+-------+-------+-------+\n";
    for (let i = 0; i < 9; ++i) {
      result += "| ";
      for (let j = 0; j < 9; ++j) {
        const value = this.getValue([i, j]);
        result += value === 0 ? "." : value;
        result += " ";
        if ((j + 1) % 3 === 0 && j < 8) {
          result += "| ";
        }
      }
      result += "|\n";
      if ((i + 1) % 3 === 0 && i < 8) {
        result += "+-------+-------+-------+\n";
      }
    }
    result += "+-------+-------+-------+\n";
    return result;
  }
}

function main() {
  const sudoku = new Sudoku();
  let inputBuffer = "";
  let rowIndex = 0;
  let colIndex = 0;

  process.stdin.setEncoding("utf8");
  process.stdin.on("data", (chunk) => {
    inputBuffer += chunk;

    while (inputBuffer.length > 0 && rowIndex < 9) {
      const char = inputBuffer[0];
      inputBuffer = inputBuffer.slice(1);
      if (/\s/.test(char)) {
        continue;
      }
      if (/[1-9]/.test(char)) {
        const value = parseInt(char, 10);
        sudoku.setValue([rowIndex, colIndex], value);

        ++colIndex;
        if (colIndex === 9) {
          colIndex = 0;
          ++rowIndex;
        }
      }
    }

    if (rowIndex === 9) {
      process.stdin.pause();

      // Display and solve the puzzle
      console.log("Original:");
      console.log(sudoku.toString());

      const result = sudoku.solve();
      if (result === null) {
        console.log("No solution found.");
        process.exit(1);
      }

      console.log("Solved:");
      console.log(result.toString());
      process.exit(0);
    }
  });

  process.stdin.on("end", () => {
    if (rowIndex < 9) {
      console.error("Unexpected EOF");
      process.exit(1);
    }
  });
}

main();
